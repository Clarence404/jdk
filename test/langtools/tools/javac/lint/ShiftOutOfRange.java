/*
 * @test /nodynamiccopyright/
 * @bug 5038439
 * @summary Verify warnings about bit shifts using out-of-range shift amounts
 * @compile/ref=ShiftOutOfRange.out -XDrawDiagnostics -Xlint:lossy-conversions ShiftOutOfRange.java
 */

public class ShiftOutOfRange {

    public void shiftInt() {
        int a = 123;

        // These should generate warnings
        a = a << (byte)-1;
        a = a >> (char)-1;
        a = a >>> (short)-1;
        a <<= -1;
        a >>= -1L;          // also generates "implicit cast from long to int in compound assignment is possibly lossy"
        a >>>= (byte)-1;

        // These should not generate warnings
        a = a << (byte)16;
        a = a >> (char)16;
        a = a >>> (short)16;
        a <<= 16;
        a >>= (long)16;     // also generates "implicit cast from long to int in compound assignment is possibly lossy"
        a >>>= (byte)16;

        // These should generate warnings
        a = a << (byte)32;
        a = a >> (char)32;
        a = a >>> (short)32;
        a <<= 32;
        a >>= (long)32;     // also generates "implicit cast from long to int in compound assignment is possibly lossy"
        a >>>= (byte)32;
    }

    public void shiftLong() {
        long a = 123;

        // These should generate warnings
        a = a << (byte)-1;
        a = a >> (char)-1;
        a = a >>> (short)-1;
        a <<= -1;
        a >>= -1L;
        a >>>= (byte)-1;

        // These should not generate warnings
        a = a << (byte)32;
        a = a >> (char)32;
        a = a >>> (short)32;
        a <<= 32;
        a >>= (long)32;
        a >>>= (byte)32;

        // These should generate warnings
        a = a << (byte)64;
        a = a >> (char)64;
        a = a >>> (short)64;
        a <<= 64;
        a >>= (long)64;
        a >>>= (byte)64;
    }
}
