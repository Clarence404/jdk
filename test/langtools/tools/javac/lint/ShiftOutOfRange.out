ShiftOutOfRange.java:14:18: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:15:18: compiler.warn.bit.shift.out.of.range: int, 65535, 31
ShiftOutOfRange.java:16:19: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:17:15: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:18:15: compiler.warn.possible.loss.of.precision: long, int
ShiftOutOfRange.java:18:15: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:19:16: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:26:15: compiler.warn.possible.loss.of.precision: long, int
ShiftOutOfRange.java:30:18: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:31:18: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:32:19: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:33:15: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:34:15: compiler.warn.possible.loss.of.precision: long, int
ShiftOutOfRange.java:34:15: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:35:16: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:42:18: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:43:18: compiler.warn.bit.shift.out.of.range: long, 65535, 63
ShiftOutOfRange.java:44:19: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:45:15: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:46:15: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:47:16: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:58:18: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:59:18: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:60:19: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:61:15: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:62:15: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:63:16: compiler.warn.bit.shift.out.of.range: long, 64, 0
27 warnings
